### PR TITLE
Add the "folder structure" chapter and modify the "extension" chapter

### PR DESCRIPTION
**Pull request by lionel-m**
This is the same as #345 but without adding a new chapter.

### CODE REVIEW

**Comment by aschempp** [comment]
Wow, this is great!

Two things:
- Contao 3 extensions can also be installed with Composer, if the developer said them to be compatible with Contao 4
- I would suggest to write the full class name instead of a use statement in the kernel.

**Comment by lionel-m** [comment]
> Contao 3 extensions can also be installed with Composer, if the developer said them to be compatible with Contao 4

Ok I will separate into two sub-chapter "with Composer" and "manually".

> I would suggest to write the full class name instead of a use statement in the kernel.

Can I have an example please?

**Comment by aschempp** [comment]
> Can I have an example please?

``` php
public function registerBundles()
{
    $bundles = [
        new Symfony\Bundle\FrameworkBundle\FrameworkBundle(),
        new Contao\CoreBundle\ContaoCoreBundle(),
        new Contao\CoreBundle\HttpKernel\Bundle\ContaoModuleBundle('myExtensionName', $this->getRootDir()),
    ];

    return $bundles;
}
```

**Comment by lionel-m** [comment]
Thanks!

**Comment by lionel-m** [comment]
> Contao 3 extensions can also be installed with Composer, if the developer said them to be compatible with Contao 4

I added this part in  7bc9f0f8605bc9f36395a5834e07d4e516825c2b.
Should we add a line in the `app/AppKernel.php` file if we install an extension with Composer?

**Comment by lionel-m** [comment]
@aschempp Could you add a new branch for the 4.2 please

File: manual/en/05-system-administration/extensions.md [hunk 1]
```diff
@@ -1,58 +1,108 @@
...
 
-Contao will automatically download and install the extension and update the
-database if necessary.
 
-![](images/extension-install.jpg)
+#### With Composer
 
+An extension that can be installed via Composer can be found through its main
+repository [Packagist][2]. A name of an extension is divided into two parts.
+The first part is the name of the project owner and the second the extension
```
Comment (aschempp) [inline manual/en/05-system-administration/extensions.md]: _project owner_ is usually called _vendor_

File: manual/en/05-system-administration/extensions.md [hunk 2]
```diff
@@ -1,58 +1,108 @@
...
 
-### Extension manager
+Dependencies (in our case an extension) are described in a file named
+`composer.json` which is located in the root folder of your Contao installation.
 
-The "extension manager" module allows you to update and uninstall extensions. It
-automatically checks for updates and notifies you if a new version is available.
-Many extensions also include links to an online manual and/or forum thread where
-you can get support.
+This is the first step you need to do. Open the `composer.json` file and add the
```
Comment (aschempp) [inline manual/en/05-system-administration/extensions.md]: The whole chapter can probably be simplified a lot. One should simply run `php composer.phar require companyName/extensionName` and let Composer handle the rest (update `composer.json` and determining the best version).

File: manual/en/05-system-administration/extensions.md [hunk 3]
```diff
@@ -1,58 +1,108 @@
...
-the database if necessary. Note that this action cannot be undone and the tables
-cannot be restored!
+You must also enter the version you want to use. The `~` [operator][3] means you
+want the latest version of `1.*`.
 
-![](images/extension-uninstall.jpg)
+Run the command `php composer.phar update companyName/extensionName` in your
+command-line interface to start the installation.
 
+After the installation, you can clear the cache with the following command:
```
Comment (aschempp) [inline manual/en/05-system-administration/extensions.md]: In a Contao 4 standard edition, this is automatically done by Composer (using the scripts defined in your root `composer.json`)

File: manual/en/05-system-administration/extensions.md [hunk 4]
```diff
@@ -1,58 +1,108 @@
...
+them to the `system/modules` folder. If the extension has public files, you must
+generate a [symbolic link][4] with the command `app/console contao:symlinks` in
+your command-line interface. Then you must register your extension in
+`app/AppKernel.php` so that it can be taken into account by the system (see
+below). Finally, check the database with the [Contao install tool][5].
+
+
+##### AppKernel.php
+
+Instantiate the `ContaoModuleBundle` class. The first parameter is the name of
```
Comment (aschempp) [inline manual/en/05-system-administration/extensions.md]: Maybe we can reuse the examples from http://symfony.com/doc/current/cookbook/bundles/best_practices.html#installation-instructions for how to adjust the `AppKernel`?

File: manual/en/05-system-administration/extensions.md (hunk 1 above)
Comment (lionel-m) [inline manual/en/05-system-administration/extensions.md]: I kept "project owner" in parentheses because everyone does not necessarily know this term.

Comment (lionel-m) [inline manual/en/05-system-administration/extensions.md]: Like me for example ;-)

File: manual/en/05-system-administration/extensions.md (hunk 2 above)
Comment (lionel-m) [inline manual/en/05-system-administration/extensions.md]: Thank you, it's easier that way. Modified in e3ca99887bc4ef975a669cd8b9c63d733a97985a

File: manual/en/05-system-administration/extensions.md (hunk 4 above)
Comment (lionel-m) [inline manual/en/05-system-administration/extensions.md]: It's done in 7b553cd74fa5b65f530c0e5af4949908211f39be

**Comment by lionel-m** [comment]
@aschempp I made all the changes.

**Comment by aschempp** [comment]
> Public files of each bundle are located in the `web/bundles` folder through symlinks. These can be regenerated from the back end under maintenance.

Not sure if that's correct. @leofeyer do we call `assets:install` from the backend or just `contao:symlinks`?

**Comment by aschempp** [comment]
The part about installing extensions is a bit confusing to me. It explains about bundles, but what we're installing in the example seems to be old Contao 3 extensions. Are you intentionally not mentioning how to install a real bundle? Maybe we should be more clear about the differences?

**Comment by lionel-m** [comment]
> Are you intentionally not mentioning how to install a real bundle? Maybe we should be more clear about the differences?

No it's not intentional. I will improve this point... Add the installation of a real bundle and make the difference between the two types of extension.

**Comment by leofeyer** [comment]
Just `contao:symlinks`.

**Comment by lionel-m** [comment]
> No it's not intentional. I will improve this point... Add the installation of a real bundle and make the difference between the two types of extension.

@aschempp it's better now WDYT?

**Comment by aschempp** [comment]
Great progress! For a Contao 3 extension to be installed in Contao 4, a developer has to mark it as compatible. Therefore the warning about compatibility should probably go into the "manual installation" section, what do you think?

We should also add a link to Packagist that lists all Contao things, they currently look like this:
https://packagist.org/search/?q=&type=contao-bundle (Contao 4 bundles) or
https://packagist.org/search/?q=&type=contao-module (Contao 3 extensions)

**Comment by lionel-m** [comment]
I removed the warning and improved the existing chapter (Contao modules) on this subject in c066a7c9f679e44910a1eb5c26d2c9e927aed069.

**Comment by lionel-m** [comment]
> We should also add a link to Packagist that lists all Contao things, they currently look like this:
> https://packagist.org/search/?q=&type=contao-bundle (Contao 4 bundles) or
> https://packagist.org/search/?q=&type=contao-module (Contao 3 extensions)

I added a new sub-chapter (Finding extensions or bundles).

**Comment by lionel-m** [comment]
What do you think about changing the chapter name: "Extensions and bundles" instead of "Extensions" only?

**Comment by aschempp** [comment]
I like it 😎 
We should maybe add more information about the different types of extensions, and why you should use one or the other. I could create a pull request for you with my ideas, ok?

**Comment by lionel-m** [comment]
> I could create a pull request for you with my ideas, ok?

Yes thx 👍

**Comment by lionel-m** [comment]
@aschempp I recreate this PR for the 4.2 branch with the last changes?

UPDATE: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request 👍

**Comment by aschempp** [comment]
Reviewed this again. I wonder if the folder structure belongs to the chapter "installing Contao"? Not really relevant for a regular user, is it? Maybe we should move this to a cookbook chapter, what do you think?

**Comment by lionel-m** [comment]
> Reviewed this again. I wonder if the folder structure belongs to the chapter "installing Contao"? Not really relevant for a regular user, is it? Maybe we should move this to a cookbook chapter, what do you think?

OK I will create a chapter with this part in the cookbook and recreating this PR that will target the 4.2 branch.

**Comment by lionel-m** [comment]
Closed in favor of #424 and #425.

**Pull request closed without merging**